stablehlo/transforms/StablehloWrapInComposite.cpp

```diff
@@ -0,0 +1,216 @@
+  mlir::SymbolTable symbol_table(module);
+  implFunc.setPrivate();
+  symbol_table.insert(implFunc);
+
+  return implFunc;
+}
+
+// A ConversionPattern that matches any operation and rewrites it as a
+// stablehlo::CompositeOp. The original operation's functionality is
+// encapsulated within a newly created private function.
+class ConvertGenericOp : public ConversionPattern {
+ public:
+  explicit ConvertGenericOp(MLIRContext* ctx)
+      : ConversionPattern(MatchAnyOpTypeTag(), /*benefit=*/1, ctx) {}
+
+  LogicalResult matchAndRewrite(
+      Operation* op, ArrayRef<Value> operands,
+      ConversionPatternRewriter& rewriter) const override {
+    // get the enclosing module
+    auto module = op->getParentOfType<ModuleOp>();
+    MLIRContext* context = op->getContext();
+    if (module == nullptr) {
+      return rewriter.notifyMatchFailure(op, "Failed to find enclosing module");
+    }
+    auto implFunc = buildStableHLOCompositeImplFunc(module, op);
+    auto name = op->getName().getStringRef();
+
+    llvm::SmallVector<mlir::Value> compositeOperands(op->operand_begin(),
+                                                     op->operand_end());
+    auto compositeOp = rewriter.create<stablehlo::CompositeOp>(
+        op->getLoc(), op->getResultTypes(), compositeOperands, name,
+        DictionaryAttr::get(context, op->getAttrs()), implFunc.getSymName());
+    rewriter.replaceOp(op, compositeOp.getResults());
+    return success();
+  }
+};
+
+class StablehloWrapInCompositePass
+    : public impl::StablehloWrapInCompositePassBase<
+          StablehloWrapInCompositePass> {
+ public:
+  StablehloWrapInCompositePass()
+      : StablehloWrapInCompositePassBase<StablehloWrapInCompositePass>() {}
+  StablehloWrapInCompositePass(const StablehloWrapInCompositePassOptions& opts)
+      : StablehloWrapInCompositePassBase<StablehloWrapInCompositePass>(opts) {}
+  explicit StablehloWrapInCompositePass(
+      std::function<bool(Operation*)> opPredicate) {
+    this->opPredicate = opPredicate;
+  }
+  LogicalResult initialize(MLIRContext* context) override {
+    RewritePatternSet patterns_(context);
+    patterns_.add<ConvertGenericOp>(context);
+    patterns = std::move(patterns_);
+
+    if (!opPredicate) {
+      if (!opNamesOption.empty()) {
+        DenseSet<StringRef> opNames(opNamesOption.begin(), opNamesOption.end());
+        opPredicate = [opNames](Operation* op) {
+          return opNames.contains(op->getName().getStringRef());
+        };
+      } else {
+        opPredicate = [](Operation* op) { return false; };
+      }
+    }
+    return success();
+  }
+
+  void runOnOperation() override {
+    ConversionTarget target(getContext());
+    target.addLegalDialect<func::FuncDialect>();
+    target.addDynamicallyLegalDialect<stablehlo::StablehloDialect>(
+        [this](Operation* op) { return !opPredicate(op); });
+
+    Operation* op = getOperation();
+    if (failed(applyPartialConversion(op, target, std::move(patterns)))) {
```

Comment:
How does this prevent recursively wrapping an op? It looks like an op in a map is illegal?

Maybe we should use a greedy rewrite instead of a dialect rewrite and only apply the patterns once? Or potentially even just a single IR `walk`?

Reply:
After debugging this is what I got"

In the first pass of the conversion application, each illegal op gets converted to a legal `stablehlo.composite` and no further action is taken for that op. The legality check for newly created ops are based on if they are "inserted" by the `ConversionPatternRewriter`. The new private function that we inserted is based on OpBuilder.

Reply:
Applied `applyPatternsGreedily` with `GreedyRewriteStrictness::ExistingOps` strictness.